### PR TITLE
*: add metadata lock when using the plan cache

### LINKED ISSUE #51407: admin check failed with “Error 8223 (HY000): data inconsistency in table” after kill one of tikv or tikv rolling restart during adding index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1、tidb_enable_dist_task='on'
2、run sysbench
3、add index for one of table
4、kill one of tikv
5、admin check index when adding index finish

tidb logs:
[endless-ha-test-add-index-tps-7080337-1-324.tar.gz](https://github.com/pingcap/tidb/files/14442084/endless-ha-test-add-index-tps-7080337-1-324.tar.gz)

### 2. What did you expect to see? (Required)
admin check success

### 3. What did you see instead (Required)

admin check failed(Error 8223 (HY000): data inconsistency in table: sbtest1, index: index_test_1709003362736, handle: 35087534, index-values:"handle: 35087534, values: [KindString 54150009827-63531154822-50885588575-38796535820-59831497760-15874376114-47900593283-86891964691-89920570946-88503431724]" != record-values:"handle: 35087534, values: [KindString 48350368445-67037925709-53068813339-93858222535-70857817307-21035926610-73758438108-82851278774-47393726843-33708297691]")
operatorLogs:
[2024-02-27 11:09:22] ###### start adding index
alter table sbtest1 add index index_test_1709003362736 (c)
[2024-02-27 11:09:22] ###### wait for ddl job finish
[2024-02-27 11:17:30] ###### ddl job finished
select job_id, job_type, state from information_schema.ddl_jobs where query = 'alter table sbtest1 add index index_test_1709003362736 (c)'
jobId: 449, job type: add index /* ingest */, state: synced
add index done, it takes: 8m7.626657145s
[2024-02-27 11:17:30] ###### start admin check
admin check index sbtest1 index_test_1709003362736

### 4. What is your TiDB version? (Required)
./tidb-server -V
 Release Version: v8.0.0-alpha
Edition: Community
Git Commit Hash: c1b70d56de9bf063a8e8c42505cfdfd8699c1e33
Git Branch: heads/refs/tags/v8.0.0-alpha
UTC Build Time: 2024-02-26 11:45:04
GoVersion: go1.21.6
Race Enabled: false
Check Table Before Drop: false
Store: unistore
2024-02-27T10:57:36.218+0800

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #51407

Problem Summary:
When executing the prepared statement from the plan cache, tidb doesn't add the metadata lock, which may cause data-index inconsistency. 

### What changed and how does it work?
1. Collect the related table information when `PREPARE`.
2. Add metadata lock during planCachePreprocess. If the schema is changed, can't use the cache plan anymore.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
fix a bug that metadata lock is missing when using the plan cache
```

### CODE REVIEW

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: pkg/parser/model/model.go [hunk 1]
```diff
@@ -542,6 +542,8 @@ type TableInfo struct {
 
 	TTLInfo *TTLInfo `json:"ttl_info"`
 
+	TableVersion uint64 `json:"table_version"`
```
Comment (tangenta) [inline pkg/parser/model/model.go:545]: Maybe `ModificationCount` is better.

Comment (zimulala) [reply pkg/parser/model/model.go:545]: Please add a comment here. Explain the difference between "TableVersion" and "Version".

File: pkg/parser/ast/misc.go [hunk 2]
```diff
@@ -556,6 +556,7 @@ type Prepared struct {
 	StmtType      string
 	Params        []ParamMarkerExpr
 	SchemaVersion int64
+	RelateVersion map[int64]uint64
```
Comment (winoros) [inline pkg/parser/ast/misc.go:559]: Add some comments for this field.
And why the SchemaVersion is not enough for the cache key?

File: pkg/planner/core/plan_cache.go [hunk 3]
```diff
@@ -104,8 +104,32 @@ func planCachePreprocess(ctx context.Context, sctx sessionctx.Context, isNonPrep
...
+		if err != nil {
+			schemaNotMatch = true
+			continue
+		}
+		if stmt.tbls[i].Meta().TableVersion != newTbl.Meta().TableVersion {
+			schemaNotMatch = true
+		}
+		stmt.tbls[i] = newTbl
+		stmtAst.RelateVersion[newTbl.Meta().ID] = newTbl.Meta().TableVersion
+	}
```
Comment (winoros) [inline pkg/planner/core/plan_cache.go:129]: This part doesn't belong to step 2 or step 3. Please split it to a separate step. And add comments for this part.

File: pkg/planner/core/plan_cache.go [hunk 4]
```diff
@@ -104,8 +104,32 @@ func planCachePreprocess(ctx context.Context, sctx sessionctx.Context, isNonPrep
...
+				return plannererrors.ErrSchemaChanged.GenWithStack("Schema change caused error: %s", err.Error())
+			}
+			delete(stmtAst.RelateVersion, stmt.tbls[i].Meta().ID)
+			stmt.tbls[i] = tblByName
+			stmtAst.RelateVersion[tblByName.Meta().ID] = tblByName.Meta().TableVersion
+		}
+		newTbl, err := tryLockMDLAndUpdateSchemaIfNecessary(sctx.GetPlanCtx(), stmt.dbName[i], stmt.tbls[i], is)
+		if err != nil {
+			schemaNotMatch = true
+			continue
```
Comment (winoros) [inline pkg/planner/core/plan_cache.go:122]: It should be a break? Or we should iterate and fill the whole `RelateVersion`?

File: pkg/planner/core/plan_cache.go [hunk 5]
```diff
@@ -104,8 +104,32 @@ func planCachePreprocess(ctx context.Context, sctx sessionctx.Context, isNonPrep
...
+		}
+		if stmt.tbls[i].Meta().TableVersion != newTbl.Meta().TableVersion {
+			schemaNotMatch = true
```
Comment (winoros) [inline pkg/planner/core/plan_cache.go:125]: ditto

File: pkg/planner/core/plan_cache_utils.go [hunk 6]
```diff
@@ -182,6 +189,30 @@ func GeneratePlanCacheStmtWithAST(ctx context.Context, sctx sessionctx.Context,
...
+	})
+	dbName := make([]model.CIStr, 0, len(mdlMap))
+	tbls := make([]table.Table, 0, len(mdlMap))
+	for id := range mdlMap {
+		tbl, ok := is.TableByID(id)
+		if !ok {
+			logutil.BgLogger().Error("table not found in info schema", zap.Int64("tableID", id))
+			return nil, nil, 0, errors.New("table not found in info schema")
+		}
+		db, ok := is.SchemaByID(tbl.Meta().DBID)
```
Comment (winoros) [inline pkg/planner/core/plan_cache_utils.go:206]: When will we need to check the database while the table id is correctly found?

File: pkg/planner/core/plan_cache_utils.go [hunk 7]
```diff
@@ -232,6 +266,23 @@ type planCacheKey struct {
...
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return keys[i] < keys[j]
+	})
+
+	h := fnv.New64a()
+	for _, k := range keys {
+		v := m[k]
+		_, _ = fmt.Fprintf(h, "%d:%d", k, v)
```
Comment (qw4990) [inline pkg/planner/core/plan_cache_utils.go:281]: Please don't use fmt.FPrintf, it's too slow.

File: pkg/planner/core/tests/prepare/prepare_test.go [hunk 8]
```diff
@@ -1459,58 +1459,59 @@ func verifyCache(ctx context.Context, t *testing.T, tk1 *testkit.TestKit, tk2 *t
...
-	tk1.MustExec("prepare s from 'select * from t1 where id = 1 for update'")
-	stmtID, _, _, err := tk1.Session().PrepareStmt("select * from t1 where id = 1 for update")
-	require.Nil(t, err)
-	tk1.MustExec("begin pessimistic")
-
-	// Verify for the for update read
-	verifyCache(ctx, t, tk1, tk2, stmtID)
-	tk1.MustExec("rollback")
-}
+//func TestCacheHitInRc(t *testing.T) {
```
Comment (qw4990) [inline pkg/planner/core/tests/prepare/prepare_test.go:1462]: Why remove these 2 tests

File: pkg/planner/core/plan_cache_utils.go [hunk 9]
```diff
@@ -463,6 +516,7 @@ type PlanCacheStmt struct {
 
 	// below fields are for PointGet short path
 	SchemaVersion int64
+	RelateVersion map[int64]uint64
```
Comment (qw4990) [inline pkg/planner/core/plan_cache_utils.go:519]: Please add some commends

**Review by qw4990 — APPROVED** [review]

File: pkg/parser/model/model.go [hunk 10]
```diff
@@ -542,6 +542,9 @@ type TableInfo struct {
 
 	TTLInfo *TTLInfo `json:"ttl_info"`
 
+	// Revision is per table schema's version, it will be increased when the schema changed.
+	Revision uint64 `json:"mod_reversion"`
```
Comment (tangenta) [inline pkg/parser/model/model.go:546]:
```suggestion
	Revision uint64 `json:"revision"`
```

**Review by tangenta — APPROVED** [review]

**Pull request merged**